rcgen/src/string_types.rs

```diff
@@ -0,0 +1,317 @@
+/// SHOULD NOT be used for certificates for new subjects.
+///
+/// [RFC 5280]: https://datatracker.ietf.org/doc/html/rfc5280#page-25
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct TeletexString(String);
+
+impl TeletexString {
+	/// Extracts a string slice containing the entire `TeletexString`.
+	pub fn as_str(&self) -> &str {
+		&self.0
+	}
+
+	/// Extracts a string slice containing the entire `PrintableString`.
+	pub fn as_bytes(&self) -> &[u8] {
+		self.0.as_bytes()
+	}
+}
+
+impl TryFrom<&str> for TeletexString {
+	type Error = Error;
+
+	fn try_from(input: &str) -> Result<Self, Error> {
+		input.to_string().try_into()
+	}
+}
+
+impl TryFrom<String> for TeletexString {
+	type Error = Error;
+
+	fn try_from(input: String) -> Result<Self, Error> {
+		// Check all bytes are visible
+		if !input.as_bytes().iter().all(|b| 0x20 <= *b && *b <= 0x7f) {
+			return Err(Error::InvalidPrintableString);
+		}
+		Ok(Self(input.to_string()))
+	}
+}
+
+impl FromStr for TeletexString {
+	type Err = Error;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		s.try_into()
+	}
+}
+
+/// ASN.1 `BMPString` type.
+///
+/// Encodes Basic Multilingual Plane (BMP) subset of Unicode (ISO 10646),
+/// a.k.a. UCS-2.
+///
+/// `BMPString` is included for backward compatibility, [RFC 5280] say it
+/// SHOULD NOT be used for certificates for new subjects.
+///
+/// [RFC 5280]: https://datatracker.ietf.org/doc/html/rfc5280#page-25
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct BmpString(Vec<u8>);
+
+impl BmpString {
+	/// Borrow the encoded UCS-2 as bytes.
+	pub fn as_bytes(&self) -> &[u8] {
+		&self.0
+	}
+}
+
+impl TryFrom<&[u8]> for BmpString {
+	type Error = Error;
+	/// Create a new [`BmpString`] from its UCS-2 encoding.
+	fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+		if value.len() % 2 != 0 {
+			return Err(Error::InvalidBmpString);
+		}
+		for maybe_char in char::decode_utf16(
+			value
+				.chunks_exact(2)
+				.map(|chunk| u16::from_be_bytes([chunk[0], chunk[1]])),
+		) {
+			match maybe_char {
+				// Character is in the Basic Multilingual Plane
+				Ok(c) if (c as u64) < u64::from(u16::MAX) => (),
+				// Characters outside Basic Multilingual Plane or unpaired surrogates
+				_ => return Err(Error::InvalidBmpString),
+			}
+		}
+		Ok(Self(value.to_vec()))
+	}
+}
+
+impl TryFrom<&str> for BmpString {
+	type Error = Error;
+	/// Create a new [`BmpString`] from a UTF-8 string.
+	fn try_from(value: &str) -> Result<Self, Self::Error> {
+		let capacity = value
+			.len()
+			.checked_mul(2)
+			.ok_or_else(|| Error::InvalidBmpString)?;
+
+		let mut bytes = Vec::with_capacity(capacity);
+
+		for code_point in value.encode_utf16() {
+			bytes.extend(code_point.to_be_bytes());
+		}
+
+		bytes.as_slice().try_into()
+	}
+}
+
+impl FromStr for BmpString {
+	type Err = Error;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		s.try_into()
+	}
+}
+
+/// ASN.1 `BMPString` type.
```

Comment:
I think this comment headline mentions the wrong type:

```suggestion
/// ASN.1 `UniversalString` type.
```